### DISCUSSION TITLE
Help understanding the fsi flat channel input file

### DISCUSSION THREAD

**Discussion by Eilloo** [idaholab/moose#23484] [Q&A General] [answered]
Hi all,

Apologies if this is a little basic, but I am having trouble understanding exactly what is going on in the fsi module to enable the transfer of fluid pressure forces to the solid.

Based on the example input file (pasted below for convenience) and my understanding of the documentation, the interface kernels seem to be doing the following:
- Setting the fluid and solid velocities to be equal at the interface (this makes sense to me as the 'no slip' condition). 
- Setting the flux of the fluid velocity to equal the solid displacement.

The second part I am having trouble understanding, although in some ways it makes sense to me as the flux of the fluid velocity represents its displacement?

Intuitively, and based on my own reading online, I would expect something more along the lines of setting the normal force on the solid side of the interface equal to the pressure on the fluid side to enable each medium to act on the other.

My other question here is why we need the velocity source/reaction terms? I can see the effect they have by manipulating them in the simulation, but I am not clear on what they physically represent.

Can someone help explain how the coupling is achieved here, and/or point me in the direction of any documentation/papers this is based on?

Thanks!

```
[GlobalParams]
    gravity = '0 0 0'
    integrate_p_by_parts = true
    laplace = true
    convective_term = true
    transient_term = true
    pspg = true
    displacements = 'disp_x disp_y'
[]
  
[Mesh]
    [gmg]
        type = GeneratedMeshGenerator
        dim = 2
        xmin = 0
        xmax = 3.0
        ymin = 0
        ymax = 1.0
        nx = 10
        ny = 15
        elem_type = QUAD4
    []

    [subdomain1]
        type = SubdomainBoundingBoxGenerator
        bottom_left = '0.0 0.5 0'
        block_id = 1
        top_right = '3.0 1.0 0'
        input = gmg
    []

    [interface]
        type = SideSetsBetweenSubdomainsGenerator
        primary_block = '0'
        paired_block = '1'
        new_boundary = 'master0_interface'
        input = subdomain1
    []

    [break_boundary]
        type = BreakBoundaryOnSubdomainGenerator
        input = interface
    []
[]
  
[Variables]
    [./vel_x]
        block = 0
    [../]
    [./vel_y]
        block = 0
    [../]
    [./p]
        block = 0
    [../]
    [./disp_x]
    [../]
    [./disp_y]
    [../]
    [./vel_x_solid]
        block = 1
    [../]
    [./vel_y_solid]
        block = 1
    [../]
[]
  
[Kernels]
    [./vel_x_time]
        type = INSMomentumTimeDerivative
        variable = vel_x
        block = 0
        use_displaced_mesh = true
    [../]
    [./vel_y_time]
        type = INSMomentumTimeDerivative
        variable = vel_y
        block = 0
        use_displaced_mesh = true
    [../]
    [./mass]
        type = INSMass
        variable = p
        u = vel_x
        v = vel_y
        pressure = p
        block = 0
        use_displaced_mesh = true
    [../]
    [./x_momentum_space]
        type = INSMomentumLaplaceForm
        variable = vel_x
        u = vel_x
        v = vel_y
        pressure = p
        component = 0
        block = 0
        use_displaced_mesh = true
    [../]
    [./y_momentum_space]
        type = INSMomentumLaplaceForm
        variable = vel_y
        u = vel_x
        v = vel_y
        pressure = p
        component = 1
        block = 0
        use_displaced_mesh = true
    [../]
    [./vel_x_mesh]
        type = ConvectedMesh
        disp_x = disp_x
        disp_y = disp_y
        variable = vel_x
        block = 0
        use_displaced_mesh = true
    [../]
    [./vel_y_mesh]
        type = ConvectedMesh
        disp_x = disp_x
        disp_y = disp_y
        variable = vel_y
        block = 0
        use_displaced_mesh = true
    [../]
    [./disp_x_fluid]
        type = Diffusion
        variable = disp_x
        block = 0
    [../]
    [./disp_y_fluid]
        type = Diffusion
        variable = disp_y
        block = 0
    [../]
    [./accel_tensor_x]
        type = CoupledTimeDerivative
        variable = disp_x
        v = vel_x_solid
        block = 1
    [../]
    [./accel_tensor_y]
        type = CoupledTimeDerivative
        variable = disp_y
        v = vel_y_solid
        block = 1
    [../]
    [./vxs_time_derivative_term]
        type = CoupledTimeDerivative
        variable = vel_x_solid
        v = disp_x
        block = 1
    [../]
    [./vys_time_derivative_term]
        type = CoupledTimeDerivative
        variable = vel_y_solid
        v = disp_y
        block = 1
    [../]
    [./source_vxs]
        type = MatReaction
        variable = vel_x_solid
        block = 1
        mob_name = 1
    [../]
    [./source_vys]
        type = MatReaction
        variable = vel_y_solid
        block = 1
        mob_name = 1
    [../]
[]
  
[InterfaceKernels]
    [./penalty_interface_x]
        type = CoupledPenaltyInterfaceDiffusion
        variable = vel_x
        neighbor_var = disp_x
        secondary_coupled_var = vel_x_solid
        boundary = master0_interface
        penalty = 1e6
    [../]
    [./penalty_interface_y]
        type = CoupledPenaltyInterfaceDiffusion
        variable = vel_y
        neighbor_var = disp_y
        secondary_coupled_var = vel_y_solid
        boundary = master0_interface
        penalty = 1e6
    [../]
[]
  
[Modules/TensorMechanics/Master]
    [./solid_domain]
        strain = SMALL
        incremental = false
        # generate_output = 'strain_xx strain_yy strain_zz' ## Not at all necessary, but nice
      block = '1'
    [../]
[]
  
[Materials]
    [./elasticity_tensor]
        type = ComputeIsotropicElasticityTensor
        youngs_modulus = 1e2
        poissons_ratio = 0.3
        block = '1'
    [../]
    [./small_stress]
        type = ComputeLinearElasticStress
        block = 1
    [../]
    [./const]
        type = GenericConstantMaterial
        block = 0
        prop_names = 'rho mu'
        prop_values = '1  1'
    [../]
[]
  
[BCs]
    [./fluid_x_no_slip]
        type = DirichletBC
        variable = vel_x
        boundary = 'bottom'
        value = 0.0
    [../]
    [./fluid_y_no_slip]
        type = DirichletBC
        variable = vel_y
        boundary = 'bottom left_to_0'
        value = 0.0
    [../]
    [./x_inlet]
        type = FunctionDirichletBC
        variable = vel_x
        boundary = 'left_to_0'
        function = 'inlet_func'
    [../]
    [./no_disp_x]
        type = DirichletBC
        variable = disp_x
        boundary = 'bottom top left_to_1 right_to_1 left_to_0 right_to_0'
        value = 0
    [../]
    [./no_disp_y]
        type = DirichletBC
        variable = disp_y
        boundary = 'bottom top left_to_1 right_to_1 left_to_0 right_to_0'
        value = 0
    [../]
    [./solid_x_no_slip]
        type = DirichletBC
        variable = vel_x_solid
        boundary = 'top left_to_1 right_to_1'
        value = 0.0
    [../]
    [./solid_y_no_slip]
        type = DirichletBC
        variable = vel_y_solid
        boundary = 'top left_to_1 right_to_1'
        value = 0.0
    [../]
[]
  
[Preconditioning]
    [./SMP]
        type = SMP
        full = true
    [../]
[]
  
[Executioner]
    type = Transient
    num_steps = 5
    # num_steps = 60
    dt = 0.1
    dtmin = 0.1
    solve_type = 'PJFNK'
    petsc_options_iname = '-pc_type'
    petsc_options_value = 'lu'
    line_search = none
[]
  
[Outputs]
    [./out]
        type = Exodus
    [../]
[]
  
[Functions]
    [./inlet_func]
        type = ParsedFunction
        value = '(-16 * (y - 0.25)^2 + 1) * (1 + cos(t))'
    [../]
[]
```

**Comment by GiudGiud** [comment]
@cbolisetti if you could please help here

**Comment by cbolisetti** [comment]
@Eilloo can you please provide the documentation link where you got the two bullets from? That's actually not what we do, at least for acoustic fluids. For acoustics, our solution variable for the fluid is pressure and not velocity. And the solution variable for the structure is displacement. So here's what we ensure at the interface: 

- Pressure of the fluid is equal to the normal stress
- Relationship between the pressure of the fluid and acceleration of the structure as derived [here](https://mooseframework.inl.gov/modules/fsi/fsi_acoustics.html)

The link describes what we do briefly and here's a [paper](https://onlinelibrary.wiley.com/doi/10.1002/eqe.3659) that describes the FSI implementation in detail as well as a V&V for earthquake applications. 

Also, if you are ok with using acoustic fluid (small particle displacements, irrotational, inviscid), I would recommend starting with this input file instead: modules/fsi/test/tests/fsi_acoustics/1D_struc_acoustic/1D_struc_acoustic.i

**Comment by Eilloo** [comment]
@cbolisetti, 
Thanks for the response - I suspect I have misunderstood how [this](https://mooseframework.inl.gov/source/interfacekernels/CoupledPenaltyInterfaceDiffusion.html) interface kernel works. The input file above is at the bottom of the linked page too.

Here is an explanation of what I previously thought this meant:

In the example input file, the variable appears to be the fluid velocity, and the neighbour variable is solid displacement. The class description states: "This class ensures the continuity of flux of `variable` and `neighbor_var` across an interface"
I was not able to understand why we might want the flux of the fluid velocity and the flux of the solid displacement to match at the interface, and so wondered if this meant the flux of 'variable', and the _value_ of neighbour_var.
The thinking here was that the flux of the fluid velocity is its surface integral, and so this could be interpreted as a displacement.

The class description also sates that continuity is enforced between the value of the primary and secondary variables. Here, we only specify a secondary variable, which is solid velocity, so the primary variable defaults to 'variable' - in this case, the fluid velocity. This is where I got the first bullet from, thinking that it makes sense to match the fluid and solid velocities at the interface.


Unfortunately I cannot access the paper linked to; however, the other link makes me wonder if my misunderstanding was related to something else:

For some reason, I had thought the fsi module provided two routes, one with the acoustic formulation and one without. I assumed the input file above did not involve acoustics, but it sounds like the latter is maybe this is the only option?

My intention is to try fsi with some INS flows where inviscid and irrotational assumptions may not be valid. Indeed, the above example includes 'INSMomentumLaplaceForm' kernels, which include viscous terms(?), contrary to the assumptions made in the linked docs page. Again, I could be mistaken here as I am more familiar with the vector INS kernels.

**Reply by cbolisetti** [reply]
I see. Note that `variable` or `neighbor_var` are the variables that depend on what MOOSE modules or formulations your are using. Also, the FSI module has been V&V'ed for acoustic fluid domains, but I don't know how mature it is for the INS formulation. @GiudGiud or @lindsayad might be able to answer that or might know someone that would.

**Comment by lindsayad** [comment, marked as answer]
An `InterfaceKernel` is much like an `IntegratedBC`.

It might make things easier to think about the physics that `CoupledPenaltyInterfaceDiffusion` suggests, so let's imagine having just diffusion on either side of the interface. When you integrate a diffusion problem by parts, you wind up with the surface term <test, -normals * grad u>. Now imagine we have integrated by parts on the other side: <test_neighbor, normals * grad v>

Now by writing the element residual: <test, penalty * (u - v)>, what you have implicitly done is set: -normals * grad u = penalty * (u - v)

And similarly if you set the neighbor residual: <test_neighbor, -penalty * (u - v)>, then you have implicitly set: normals * grad v = -penalty * (u - v)

So what have you done? Looking at those two implicit equations you can see you've set: -normals * grad u = -normals * grad v = penalty * (u - v)

So by the first equality you have set continuity of flux (diffusive flux). And through the penalty parameter you are driving u and v to be equal to each other.

Generalizing this to the FSI problem, you are asking for continuity of *stress* between the fluid and solid, and you are driving the difference in velocities (liquid and solid) to zero. So stress for the fluid in this case (where you are using the Laplace form for the viscous term and have integrated the pressure by parts), when dotted with the normal, is component-wise: <test, p * normals(component) - mu * grad v_component * normals> where v is the velocity, and v_component is a component of the velocity. So this fluid stress will be made equal to whatever stresses exist in the solid phase.

Does that make sense?

**Reply by Eilloo** [reply]
Thanks for the explanation - I'm definitely getting there, but there's a couple of steps I still want to clarify:

Firstly, in the diffusion example, suppose we set the residual in a boundary element to <test, anything arbitrary>, are we implying that the 'anything arbitrary' is equal to 'normals*grad_u', since 'normals*grad_u' is the natural BC? 
Then, choosing penalty*(u-v) is convenient as it enables us to drive u and v to be equal (or indeed, any two variables - in the fsi case, the fluid and solid velocities).

From the input file, the continuity of stress between solid and fluid occurs when 'normals*grad_fluidVelocity' is equal to 'normals*grad_solidDisplacement'?
Firstly, does grad_solidDisplacement actually represent the stress in the solid? I'm hazy on the solid mech side of things, but I had expected this term to be proportional only to the displacement, not the gradient?
I'm happy with the above definition of fluid stress, but am still struggling with how this is implied simply by choosing fluid velocity and solid displacement as variable and neighbour variable...
Would we not want to include the pressure term, viscosity, etc. in our residual, for instance?

For what it's worth, the idea to make this work with the vector INS kernels is to accept a the velocity vector as the variable, and have the user specify a component, which is then used in calculating the residual (the displacements would be left as individual components). In the input file, you would simply have three of these kernels, all of which have the velocity vector as the variable, but each of which has taken a different component when applying 'penalty*(u-v)'. It feels slightly messy but based on the above, I think it should work?

Hopefully this all makes sense, at least in terms of explaining the bits I don't quite understand!

**Reply by lindsayad** [reply]
> Firstly, in the diffusion example, suppose we set the residual in a boundary element to <test, anything arbitrary>, are we implying that the 'anything arbitrary' is equal to 'normals_grad_u', since 'normals_grad_u' is the natural BC?

That's exactly right.

> From the input file, the continuity of stress between solid and fluid occurs when 'normals_grad_fluidVelocity' is equal to 'normals_grad_solidDisplacement'?

Continuity of stress will occur if the interface kernel element and neighbor residual statements in the interface kernel are the same (after division by the respective test functions). You can imagine that if you set the element residual to A and the neighbor residual to B, then in that case you are requiring the stress in the fluid on the boundary to be different than the stress in the solid.

> Firstly, does grad_solidDisplacement actually represent the stress in the solid? I'm hazy on the solid mech side of things, but I had expected this term to be proportional only to the displacement, not the gradient?

I don't think I said that the stress is proportional to the displacement gradient, but I do think that is the case. You can imagine that if you displaced every portion of a solid equally, then you have merely translated the solid and you would not have any internal stresses in the body.

But whether you are a solid mechanics expert or not (I am also not), luckily I don't think you need to know anything about what goes into the stress computation in the solid.

> For what it's worth, the idea to make this work with the vector INS kernels is to accept a the velocity vector as the variable, and have the user specify a component, which is then used in calculating the residual (the displacements would be left as individual components).

What would you do about the fluid velocity test functions? They are vectors. I think that if you also indexed those by component, then what you are proposing would work.

**Reply by Eilloo** [reply]
Okay, I think I'm more or less there with this now. 

I'm happy with the idea that setting the residuals of elements on each side of the boundary implies continuity of flux between u and v, for instance.

The main sticky point for me is still relating to the choice of variables and neighbour and how they specifically infer continuity of stress, as opposed to some other quantity.

So, to check: by choosing variable and neighbour variable as velocity and displacement, both of these must appear in a diffusion term in the defining equations of their respective mediums, and that diffusion term must also represent the stress in that medium?

The reason for mentioning displacement gradient is because the natural BC from our diffusion equation is normals * grad_var, so if var is displacement on the solid side, this has to represent a stress. The example you gave is a great way of thinking about this!


R.e. using interface kernels with a vector velocity variable, I hadn't thought about the test functions - good point! 
I'll make sure I take components of these too.

Looking into the code side a little more, I think there may also be issues specifying 'variable' as a vector, but 'neighbour_var' as a scalar - even if I do intend to only use one component of the vector. I believe the existing class template (understandably) doesn't allow this... perhaps I'll need to copy the interface kernel template and make some changes to this too.

**Reply by lindsayad** [reply]
> The main sticky point for me is still relating to the choice of variables and neighbour and how they specifically infer continuity of stress, as opposed to some other quantity.

So generically speaking, if you set the element and neighbor residuals to the same thing (after dividing by the respective test functions) then you will be setting the surface terms resulting from integration by parts of the variables' (`variable` and `neighbor_var`) equations equal to each other. So generically speaking the surface terms may be stress, they may be diffusive flux, they may be advective plus diffusive flux, etc.

**Reply by Eilloo** [reply]
Okay nice, this makes sense to me now.

Since we're here, if I could ask one final thing about the input file itself:
What is the purpose of the `source_vsx` and `source_vsy` kernels here? 
I would not have thought to add these, but I can see that nothing deforms if they are removed.

**Reply by lindsayad** [reply]
These are not super physical; they were just a means to incur some displacement in the solid in order to "make stuff happen". I came up with this test really quickly a few years ago just to show that MOOSE could do some kind of fluid-structure interaction (well before the `fsi` module actually existed)

**Reply by Eilloo** [reply]
Ah, I see - thanks very much for your help!

**Reply by lindsayad** [reply]
You're welcome. Great questions!
